src/geometry/euclidean.lean

```diff
@@ -33,3 +35,634 @@ normed_add_torsor V P
+begin
+  split,
+  { intro h,
+    have ha := h,
+    apply_fun abs at ha,
+    norm_num at ha,
+    rcases (abs_inner_div_norm_mul_norm_eq_one_iff x y).1 ha with ⟨hx, ⟨r, ⟨hr, hy⟩⟩⟩,
+    use [hx, r],
+    refine and.intro _ hy,
+    by_contradiction hrpos,
+    rw hy at h,
+    rw inner_div_norm_mul_norm_eq_one_of_ne_zero_of_pos_mul hx
+      (lt_of_le_of_ne' (le_of_not_lt hrpos) hr.symm) at h,
+    norm_num at h },
+  { intro h,
+    rcases h with ⟨hx, ⟨r, ⟨hr, hy⟩⟩⟩,
+    rw hy,
+    exact inner_div_norm_mul_norm_eq_neg_one_of_ne_zero_of_neg_mul hx hr }
+end
+
+/-- The undirected angle between two vectors. If either vector is 0,
+this is π/2. -/
+def angle_of_vectors (x y : V) : ℝ := (inner x y / (∥x∥ * ∥y∥)).arccos
+
+/-- The angle between two vectors does not depend on their order. -/
+lemma angle_of_vectors_comm (x y : V) : angle_of_vectors x y = angle_of_vectors y x :=
+begin
+  unfold angle_of_vectors,
+  rw [inner_comm, mul_comm]
+end
+
+/-- The angle between the negation of two vectors. -/
+@[simp] lemma angle_of_vectors_of_neg_of_neg (x y : V) :
+  angle_of_vectors (-x) (-y) = angle_of_vectors x y :=
+begin
+  unfold angle_of_vectors,
+  rw [inner_neg_neg, norm_neg, norm_neg]
+end
+
+/-- The angle between two vectors is nonnegative. -/
+lemma angle_of_vectors_nonneg (x y : V) : 0 ≤ angle_of_vectors x y :=
+real.arccos_nonneg _
+
+/-- The angle between two vectors is at most π. -/
+lemma angle_of_vectors_le_pi (x y : V) : angle_of_vectors x y ≤ real.pi :=
+real.arccos_le_pi _
+
+/-- The angle between a vector and the negation of another vector. -/
+lemma angle_of_vectors_of_neg_right (x y : V) :
+  angle_of_vectors x (-y) = real.pi - angle_of_vectors x y :=
+begin
+  unfold angle_of_vectors,
+  rw [←real.arccos_neg, norm_neg, inner_neg_right, neg_div]
+end
+
+/-- The angle between the negation of a vector and another vector. -/
+lemma angle_of_vectors_of_neg_left (x y : V) :
+  angle_of_vectors (-x) y = real.pi - angle_of_vectors x y :=
+by rw [←angle_of_vectors_of_neg_of_neg, neg_neg, angle_of_vectors_of_neg_right]
+
+/-- The angle between the zero vector and a vector. -/
+@[simp] lemma angle_of_vectors_zero_left (x : V) : angle_of_vectors 0 x = real.pi / 2 :=
+begin
+  unfold angle_of_vectors,
+  rw [inner_zero_left, zero_div, real.arccos_zero]
+end
+
+/-- The angle between a vector and the zero vector. -/
+@[simp] lemma angle_of_vectors_zero_right (x : V) : angle_of_vectors x 0 = real.pi / 2 :=
+begin
+  unfold angle_of_vectors,
+  rw [inner_zero_right, zero_div, real.arccos_zero]
+end
+
+/-- The angle between a nonzero vector and itself. -/
+@[simp] lemma angle_of_vectors_of_self_of_nonzero {x : V} (hx : x ≠ 0) :
+  angle_of_vectors x x = 0 :=
+begin
+  unfold angle_of_vectors,
+  rw [←inner_self_eq_norm_square, div_self (λ h, hx (inner_self_eq_zero.1 h)),
+      real.arccos_one]
+end
+
+/-- The angle between a nonzero vector and its negation. -/
+@[simp] lemma angle_of_vectors_of_neg_self_of_nonzero_right {x : V} (hx : x ≠ 0) :
```

Comment:
`angle_self_neg_of_nonzero`

Reply:
Renamed (and `angle_neg_self_of_nonzero` likewise).